File: android/app/src/main/kotlin/com/elysium369/meet/core/catalog/UniversalPartsCatalog.kt
package com.elysium369.meet.core.catalog

import android.content.Context
import android.content.SharedPreferences
import java.text.Normalizer
import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json

const val UNIVERSAL_PARTS_ASSET = "knowledge/catalog/pilot_hyundai_accent_verna_2005_front_end.json"

@Serializable
data class UniversalPartsPack(
    val schemaVersion: Int,
    val packId: String,
    val packVersion: String,
    val title: String,
    val publicationState: String,
    val autoPublishAllowed: Boolean,
    val disclaimer: String,
    val contentSha256: String,
    val vehicleScope: VehicleScope,
    val sourceDocuments: List<SourceDocument>,
    val parts: List<CatalogPart>,
    val procedures: List<RepairProcedure>,
    val statistics: CatalogStatistics
)

@Serializable
data class VehicleScope(
    val vehicleProfileId: String,
    val make: String,
    val models: List<String>,
    val year: Int,
    val engineDisplacementLiters: Double,
    val transmission: String,
    val bindingState: String
)

@Serializable
data class SourceDocument(
    val sourceFileName: String,
    val sourceSha256: String
)

@Serializable
data class CatalogStatistics(
    val partCount: Int,
    val procedureCount: Int,
    val verifiedTechnicalSpecificationCount: Int,
    val automaticallyPublishableCount: Int
)

@Serializable
data class CatalogPart(
    val id: String,
    val nameEs: String,
    val nameEn: String,
    val aliases: List<String>,
    val category: String,
    val system: String,
    val subsystem: String,
    val assembly: String,
    val subassembly: String? = null,
    val position: String,
    val description: String,
    val confidence: String,
    val publicationState: String,
    val compatibilityState: String,
    val compatibilityMessage: String,
    val requiredCompatibilityEvidence: List<String>,
    val technicalSpecifications: TechnicalSpecifications,
    val sourceRefs: List<CatalogSourceRef>,
    val threeDimensionalBinding: ThreeDimensionalBinding
)

@Serializable
data class TechnicalSpecifications(
    val oemNumber: String? = null,
    val torque: String? = null,
    val material: String? = null,
    val dimensions: String? = null,
    val pinout: Map<String, String>? = null
)

@Serializable
data class CatalogSourceRef(
    val sourceFileName: String,
    val sourceDocumentSha256: String,
    val sourceBlockId: String,
    val sourceTextHash: String,
    val sectionPath: List<String>,
    val sourceKind: String,
    val reviewStatus: String
)

@Serializable
data class ThreeDimensionalBinding(
    val sceneId: String,
    val nodeId: String,
    val visualAuthority: String,
    val isDimensionalModel: Boolean
)

@Serializable
data class RepairProcedure(
    val id: String,
    val title: String,
    val targetPartIds: List<String>,
    val publicationState: String,
    val executionPolicy: String,
    val sourceRefs: List<CatalogSourceRef>,
    val difficulty: String,
    val safetyLevel: String,
    val steps: List<RepairStep>
)

@Serializable
data class RepairStep(
    val id: String,
    val order: Int,
    val title: String,
    val instruction: String,
    val warning: String? = null,
    val tools: List<String>,
    val requiredEvidence: List<String>,
    val targetPartId: String,
    val targetNodeId: String,
    val animationAction: String,
    val completionGate: String,
    val technicalValue: String? = null,
    val technicalValueMessage: String? = null
)

object UniversalPartsCatalogParser {
    private val json = Json { ignoreUnknownKeys = false }

    fun decode(raw: String): UniversalPartsPack {
        val pack = json.decodeFromString<UniversalPartsPack>(raw)
        val errors = UniversalPartsCatalogValidator.validate(pack)
        require(errors.isEmpty()) { errors.joinToString(prefix = "Invalid universal parts pack: ", separator = "; ") }
        return pack
    }
}

object UniversalPartsCatalogValidator {
    fun validate(pack: UniversalPartsPack): List<String> = buildList {
        if (pack.schemaVersion != 1) add("unsupported schemaVersion")
        if (pack.packId != "pilot_hyundai_accent_verna_2005_front_end") add("unexpected packId")
        if (pack.publicationState != "REVIEW_REQUIRED" || pack.autoPublishAllowed) add("unsafe publication state")
        if (pack.parts.size < 50) add("at least 50 parts are required")
        if (pack.statistics.partCount != pack.parts.size) add("part count mismatch")
        if (pack.statistics.procedureCount != pack.procedures.size) add("procedure count mismatch")
        if (pack.statistics.verifiedTechnicalSpecificationCount != 0) add("pilot cannot contain verified specifications")

        val ids = pack.parts.map { it.id }
        if (ids.distinct().size != ids.size) add("duplicate part id")
        val knownIds = ids.toSet()
        pack.parts.forEach { part ->
            if (part.sourceRefs.isEmpty()) add("${part.id} has no source")
            if (part.confidence != "UNVERIFIED") add("${part.id} has unsafe confidence")
            if (part.publicationState != "REVIEW_REQUIRED") add("${part.id} has unsafe publication state")
            if (part.compatibilityState != "REQUIRES_VERIFICATION") add("${part.id} has unsafe compatibility")
            if (part.technicalSpecifications.hasAnyValue()) add("${part.id} exposes unverified specification")
            if (part.threeDimensionalBinding.nodeId != part.id) add("${part.id} has broken 3D binding")
            if (part.threeDimensionalBinding.visualAuthority != "GENERIC_SCHEMATIC") add("${part.id} overstates visual authority")
            if (part.threeDimensionalBinding.isDimensionalModel) add("${part.id} cannot be dimensional")
        }
        if (pack.procedures.size < 3) add("at least three procedures are required")
        val stepIds = mutableSetOf<String>()
        pack.procedures.forEach { procedure ->
            if (procedure.executionPolicy != "TRAINING_ONLY_REVIEW_REQUIRED") add("${procedure.id} has unsafe execution policy")
            procedure.targetPartIds.filterNot(knownIds::contains).forEach { add("${procedure.id} targets unknown part $it") }
            procedure.steps.forEach { step ->
                if (!stepIds.add(step.id)) add("duplicate step id ${step.id}")
                if (step.targetPartId !in knownIds || step.targetNodeId !in knownIds) add("${step.id} has broken target")
                if (step.completionGate == "VERIFIED_TORQUE_REQUIRED" && step.technicalValue != null) {
                    add("${step.id} exposes unverified torque")
                }
            }
        }
    }

    private fun TechnicalSpecifications.hasAnyValue(): Boolean =
        oemNumber != null || torque != null || material != null || dimensions != null || pinout != null
}

class UniversalPartsCatalogRepository(
    context: Context,
    private val assetPath: String = UNIVERSAL_PARTS_ASSET
) {
    private val appContext = context.applicationContext
    @Volatile private var cached: UniversalPartsPack? = null

    fun load(): UniversalPartsPack = cached ?: synchronized(this) {
        cached ?: appContext.assets.open(assetPath).bufferedReader().use { reader ->
            UniversalPartsCatalogParser.decode(reader.readText())
        }.also { cached = it }
    }
}

data class CompatibilityAssessment(
    val state: String,
    val message: String,
    val missingEvidence: List<String>
)

object UniversalPartsCatalogEngine {
    fun search(parts: List<CatalogPart>, query: String): List<CatalogPart> {
        val needle = query.normalized()
        if (needle.isBlank()) return parts
        return parts.filter { part ->
            listOf(part.nameEs, part.nameEn, part.category, part.system, part.subsystem, part.assembly)
                .plus(part.aliases)
                .any { it.normalized().contains(needle) }
        }
    }

    fun assessCompatibility(part: CatalogPart, evidenceIds: Set<String>): CompatibilityAssessment {
        val missing = part.requiredCompatibilityEvidence.filterNot(evidenceIds::contains)
        return CompatibilityAssessment(
            state = "REQUIRES_VERIFICATION",
            message = part.compatibilityMessage,
            missingEvidence = missing
        )
    }

    fun canCompleteStep(
        step: RepairStep,
        evidenceIds: Set<String> = emptySet(),
        hasVerifiedTechnicalClaim: Boolean = false
    ): StepGateResult {
        if (step.completionGate == "VERIFIED_TORQUE_REQUIRED" && !hasVerifiedTechnicalClaim) {
            return StepGateResult(false, "Torque no confirmado para esta variante. Adjunte una fuente tecnica verificada.")
        }
        if (step.completionGate != "MANUAL_CONFIRMATION") {
            val missing = step.requiredEvidence.filterNot(evidenceIds::contains)
            if (missing.isNotEmpty()) return StepGateResult(false, "Falta evidencia requerida: ${missing.joinToString()}")
        }
        return StepGateResult(true, null)
    }

    private fun String.normalized(): String = Normalizer.normalize(this, Normalizer.Form.NFD)
        .replace(Regex("\\p{Mn}+"), "")
        .lowercase()
        .trim()
}

data class StepGateResult(val allowed: Boolean, val reason: String?)

@Serializable
data class RepairProgress(
    val procedureId: String,
    val packVersion: String,
    val state: String = "NOT_STARTED",
    val completedStepIds: Set<String> = emptySet(),
    val blockedStepId: String? = null,
    val updatedAtEpochMillis: Long = 0L
)

data class ProgressTransition(val progress: RepairProgress, val reason: String? = null)

object RepairProgressEngine {
    fun toggleStep(
        progress: RepairProgress,
        procedure: RepairProcedure,
        stepId: String,
        evidenceIds: Set<String> = emptySet(),
        hasVerifiedTechnicalClaim: Boolean = false,
        nowEpochMillis: Long = System.currentTimeMillis()
    ): ProgressTransition {
        val step = procedure.steps.firstOrNull { it.id == stepId }
            ?: return ProgressTransition(progress, "Paso desconocido")
        if (stepId in progress.completedStepIds) {
            val remaining = progress.completedStepIds - stepId
            return ProgressTransition(
                progress.copy(
                    state = if (remaining.isEmpty()) "NOT_STARTED" else "IN_PROGRESS",
                    completedStepIds = remaining,
                    blockedStepId = null,
                    updatedAtEpochMillis = nowEpochMillis
                )
            )
        }
        val gate = UniversalPartsCatalogEngine.canCompleteStep(step, evidenceIds, hasVerifiedTechnicalClaim)
        if (!gate.allowed) {
            return ProgressTransition(
                progress.copy(state = "BLOCKED", blockedStepId = stepId, updatedAtEpochMillis = nowEpochMillis),
                gate.reason
            )
        }
        val completed = progress.completedStepIds + stepId
        return ProgressTransition(
            progress.copy(
                state = if (completed.size == procedure.steps.size) "COMPLETED" else "IN_PROGRESS",
                completedStepIds = completed,
                blockedStepId = null,
                updatedAtEpochMillis = nowEpochMillis
            )
        )
    }
}

class RepairProgressStore(context: Context) {
    private val preferences: SharedPreferences =
        context.applicationContext.getSharedPreferences("meet_repair_progress", Context.MODE_PRIVATE)
    private val json = Json { ignoreUnknownKeys = true }

    fun load(procedureId: String, packVersion: String): RepairProgress {
        val fallback = RepairProgress(procedureId = procedureId, packVersion = packVersion)
        val raw = preferences.getString(key(procedureId, packVersion), null) ?: return fallback
        return runCatching { json.decodeFromString<RepairProgress>(raw) }
            .getOrNull()
            ?.takeIf { it.procedureId == procedureId && it.packVersion == packVersion }
            ?: fallback
    }

    fun save(progress: RepairProgress) {
        preferences.edit()
            .putString(key(progress.procedureId, progress.packVersion), json.encodeToString(progress))
            .apply()
    }

    private fun key(procedureId: String, packVersion: String): String = "$packVersion:$procedureId"
}

File: services/universalPartsCatalog.ts
import pilotPackJson from '../android/app/src/main/assets/knowledge/catalog/pilot_hyundai_accent_verna_2005_front_end.json';
import type { DetailedPart, GuidedRepairProcedure, RepairStep3D } from '../types';

export type CatalogPublicationState = 'REVIEW_REQUIRED' | 'PUBLISHED' | 'REJECTED';
export type CatalogCompatibilityState = 'REQUIRES_VERIFICATION' | 'PROBABLE' | 'EXACT';
export type RepairProgressState = 'NOT_STARTED' | 'IN_PROGRESS' | 'BLOCKED' | 'COMPLETED';

interface PilotSourceRef {
  sourceFileName: string;
  sourceDocumentSha256: string;
  sourceBlockId: string;
  sourceTextHash: string;
  sectionPath: string[];
  reviewStatus: string;
}

interface PilotPart {
  id: string;
  nameEs: string;
  nameEn: string;
  aliases: string[];
  category: string;
  system: string;
  subsystem: string;
  assembly: string;
  subassembly: string | null;
  position: DetailedPart['position'];
  description: string;
  confidence: 'UNVERIFIED';
  publicationState: CatalogPublicationState;
  compatibilityState: CatalogCompatibilityState;
  compatibilityMessage: string;
  requiredCompatibilityEvidence: string[];
  technicalSpecifications: {
    oemNumber: string | null;
    torque: string | null;
    material: string | null;
    dimensions: string | null;
    pinout: Record<string, string> | null;
  };
  sourceRefs: PilotSourceRef[];
  threeDimensionalBinding: {
    sceneId: string;
    nodeId: string;
    visualAuthority: 'GENERIC_SCHEMATIC';
    isDimensionalModel: false;
  };
}

interface PilotStep {
  id: string;
  order: number;
  title: string;
  instruction: string;
  warning: string | null;
  tools: string[];
  requiredEvidence: string[];
  targetPartId: string;
  targetNodeId: string;
  animationAction: 'HIGHLIGHT' | 'ISOLATE' | 'REMOVE' | 'INSTALL' | 'RESET';
  completionGate: NonNullable<RepairStep3D['completion_gate']>;
  technicalValue: string | null;
  technicalValueMessage: string | null;
}

interface PilotProcedure {
  id: string;
  title: string;
  difficulty: 'INTERMEDIATE' | 'ADVANCED';
  safetyLevel: GuidedRepairProcedure['safety_level'];
  targetPartIds: string[];
  publicationState: 'REVIEW_REQUIRED';
  executionPolicy: 'TRAINING_ONLY_REVIEW_REQUIRED';
  steps: PilotStep[];
}

interface PilotPack {
  schemaVersion: number;
  packId: string;
  packVersion: string;
  publicationState: 'REVIEW_REQUIRED';
  autoPublishAllowed: false;
  contentSha256: string;
  disclaimer: string;
  parts: PilotPart[];
  procedures: PilotProcedure[];
}

export interface RepairProgress {
  procedureId: string;
  packVersion: string;
  state: RepairProgressState;
  completedStepIds: string[];
  blockedStepId: string | null;
  updatedAt: string;
}

const pack = pilotPackJson as unknown as PilotPack;

const localizedLabels: Record<string, string> = {
  Suspension: 'Suspensión',
  'Suspension / Chasis': 'Suspensión / Chasis',
  'Direccion / Suspension': 'Dirección / Suspensión',
  Direccion: 'Dirección',
  Electrico: 'Eléctrico',
  Transmision: 'Transmisión',
  'Transmision / Tren motriz': 'Transmisión / Tren motriz',
  Fijacion: 'Fijación',
  Rotulas: 'Rótulas',
  'Rotulas de suspension': 'Rótulas de suspensión',
};

const localize = (value: string): string => localizedLabels[value] ?? value;

function assertPilotPack(value: PilotPack): void {
  if (value.schemaVersion !== 1 || value.packId !== 'pilot_hyundai_accent_verna_2005_front_end') {
    throw new Error('Unsupported MEET parts catalog pack');
  }
  if (value.autoPublishAllowed || value.publicationState !== 'REVIEW_REQUIRED') {
    throw new Error('Pilot catalog cannot be auto-published');
  }
  if (value.parts.length < 50 || value.procedures.length < 3) {
    throw new Error('Pilot catalog is incomplete');
  }
  const ids = new Set<string>();
  for (const part of value.parts) {
    if (ids.has(part.id)) throw new Error(`Duplicate part id: ${part.id}`);
    ids.add(part.id);
    if (!part.sourceRefs.length) throw new Error(`Missing source for ${part.id}`);
    if (part.compatibilityState !== 'REQUIRES_VERIFICATION' || part.confidence !== 'UNVERIFIED') {
      throw new Error(`Unsafe promotion for ${part.id}`);
    }
    if (Object.values(part.technicalSpecifications).some(item => item !== null)) {
      throw new Error(`Unverified technical specification for ${part.id}`);
    }
    if (part.threeDimensionalBinding.nodeId !== part.id) {
      throw new Error(`Broken 3D binding for ${part.id}`);
    }
  }
  for (const procedure of value.procedures) {
    for (const step of procedure.steps) {
      if (!ids.has(step.targetPartId) || !ids.has(step.targetNodeId)) {
        throw new Error(`Broken procedure target: ${step.id}`);
      }
      if (step.completionGate === 'VERIFIED_TORQUE_REQUIRED' && step.technicalValue !== null) {
        throw new Error(`Unverified torque exposed by ${step.id}`);
      }
    }
  }
}

assertPilotPack(pack);

const animationMap: Record<PilotStep['animationAction'], RepairStep3D['animation_action']> = {
  HIGHLIGHT: 'NONE',
  ISOLATE: 'EXPLODE',
  REMOVE: 'TRANSLATE_X',
  INSTALL: 'TRANSLATE_X',
  RESET: 'NONE',
};

function stepType(step: PilotStep): RepairStep3D['type'] {
  if (step.completionGate === 'VERIFIED_TORQUE_REQUIRED') return 'TORQUE';
  if (step.completionGate === 'ALIGNMENT_EVIDENCE_REQUIRED') return 'ALIGN';
  if (step.animationAction === 'REMOVE') return 'DISASSEMBLE';
  if (step.animationAction === 'INSTALL') return 'ASSEMBLE';
  return 'INSPECT';
}

export const SOURCE_BACKED_PARTS_CATALOG: DetailedPart[] = pack.parts.map(part => ({
  id: part.id,
  name: part.nameEs,
  aliases: [...part.aliases, part.nameEn],
  category: localize(part.category),
  system: localize(part.system),
  subsystem: localize(part.subsystem),
  assembly: localize(part.assembly),
  subassembly: part.subassembly ? localize(part.subassembly) : undefined,
  description: part.description,
  position: part.position,
  specification: {
    oem_number: part.technicalSpecifications.oemNumber ?? '',
    equivalent_numbers: [],
    dimensions: part.technicalSpecifications.dimensions ?? '',
    material: part.technicalSpecifications.material ?? '',
    torque_nm: part.technicalSpecifications.torque ?? undefined,
    pinout: part.technicalSpecifications.pinout ?? undefined,
  },
  symptoms: [],
  related_dtcs: [],
  confidence_level: 'UNCONFIRMED',
  publication_state: part.publicationState,
  compatibility_state: part.compatibilityState,
  compatibility_message: part.compatibilityMessage,
  required_compatibility_evidence: part.requiredCompatibilityEvidence,
  visual_authority: part.threeDimensionalBinding.visualAuthority,
  source_refs: part.sourceRefs.map(ref => ({
    source_file_name: ref.sourceFileName,
    source_document_sha256: ref.sourceDocumentSha256,
    source_block_id: ref.sourceBlockId,
    source_text_hash: ref.sourceTextHash,
    section_path: ref.sectionPath,
    review_status: ref.reviewStatus,
  })),
}));

export const SOURCE_BACKED_REPAIR_PROCEDURES: GuidedRepairProcedure[] = pack.procedures.map(procedure => ({
  id: procedure.id,
  title: procedure.title,
  vehicle_applicability: 'Perfil piloto; compatibilidad exacta pendiente de VIN/OEM y fuente aprobada.',
  estimated_duration_min: 0,
  difficulty: procedure.difficulty === 'ADVANCED' ? 'HARD' : 'MEDIUM',
  safety_level: procedure.safetyLevel,
  prerequisites: ['Validar el vehiculo, la pieza y el procedimiento aplicable antes de intervenir.'],
  steps: procedure.steps.map(step => ({
    id: step.id,
    order: step.order,
    title: step.title,
    description: step.instruction,
    type: stepType(step),
    target_node_id: step.targetNodeId,
    animation_action: animationMap[step.animationAction],
    required_tools: step.tools,
    torque_spec: step.completionGate === 'VERIFIED_TORQUE_REQUIRED'
      ? (step.technicalValueMessage ?? 'No confirmado para esta variante')
      : undefined,
    warning_notes: step.warning ?? undefined,
    completion_gate: step.completionGate,
    required_evidence: step.requiredEvidence,
    technical_value_message: step.technicalValueMessage ?? undefined,
  })),
  final_verification: procedure.id.startsWith('verify_')
    ? ['Alineación documentada.', 'Prueba funcional controlada.', 'Reinspección final registrada.']
    : ['Continuar con el procedimiento de verificación posterior antes de cerrar el trabajo.'],
  publication_state: procedure.publicationState,
  execution_policy: procedure.executionPolicy,
}));

export const SOURCE_BACKED_PACK_META = Object.freeze({
  packId: pack.packId,
  packVersion: pack.packVersion,
  contentSha256: pack.contentSha256,
  publicationState: pack.publicationState,
  disclaimer: pack.disclaimer,
});

function normalized(value: string): string {
  return value.normalize('NFD').replace(/[\u0300-\u036f]/g, '').toLowerCase().trim();
}

export function searchSourceBackedParts(query: string): DetailedPart[] {
  const term = normalized(query);
  if (!term) return SOURCE_BACKED_PARTS_CATALOG;
  return SOURCE_BACKED_PARTS_CATALOG.filter(part =>
    [part.name, ...part.aliases, part.category, part.system, part.subsystem, part.assembly]
      .some(value => normalized(value).includes(term))
  );
}

export function canCompleteRepairStep(
  step: RepairStep3D,
  evidenceIds: ReadonlySet<string> = new Set(),
  hasVerifiedTechnicalClaim = false,
): { allowed: boolean; reason: string | null } {
  if (step.completion_gate === 'VERIFIED_TORQUE_REQUIRED' && !hasVerifiedTechnicalClaim) {
    return { allowed: false, reason: 'Torque no confirmado para esta variante. Adjunte una fuente técnica verificada.' };
  }
  if (step.completion_gate !== 'MANUAL_CONFIRMATION') {
    const missing = (step.required_evidence ?? []).filter(item => !evidenceIds.has(item));
    if (missing.length > 0) {
      return { allowed: false, reason: `Falta evidencia requerida: ${missing.join(', ')}` };
    }
  }
  return { allowed: true, reason: null };
}

const progressKey = (procedureId: string): string =>
  `meet.repair-progress.${pack.packVersion}.${procedureId}`;

export function loadRepairProgress(procedureId: string): RepairProgress {
  const empty: RepairProgress = {
    procedureId,
    packVersion: pack.packVersion,
    state: 'NOT_STARTED',
    completedStepIds: [],
    blockedStepId: null,
    updatedAt: new Date(0).toISOString(),
  };
  if (typeof window === 'undefined') return empty;
  try {
    const stored = window.localStorage.getItem(progressKey(procedureId));
    if (!stored) return empty;
    const parsed = JSON.parse(stored) as RepairProgress;
    return parsed.procedureId === procedureId && parsed.packVersion === pack.packVersion ? parsed : empty;
  } catch {
    return empty;
  }
}

export function saveRepairProgress(progress: RepairProgress): void {
  if (typeof window === 'undefined') return;
  try {
    window.localStorage.setItem(progressKey(progress.procedureId), JSON.stringify(progress));
  } catch {
    // Progress remains usable in memory when storage is unavailable or full.
  }
}
